src/adapter/etl-adapter-google-sheet/src/Flow/ETL/Adapter/GoogleSheet/GoogleSheetLoader.php

```diff
@@ -0,0 +1,138 @@
+use Google\Service\Sheets;
+use Google\Service\Sheets\ValueRange;
+
+final class GoogleSheetLoader implements Loader, Loader\Closure
+{
+    private ?int $columnsCount = null;
+
+    private string $dateFormat = 'Y-m-d';
+
+    private string $dateTimeFormat = 'Y-m-d H:i:s';
+
+    private ValueInputOption $inputOption = ValueInputOption::USER_ENTERED;
+
+    private int $loadedRows = 0;
+
+    private string $timeFormat = 'H:i';
+
+    private bool $withHeader = true;
+
+    public function __construct(
+        private readonly Sheets $service,
+        private readonly string $spreadsheetId,
+        private readonly string $sheetName,
+    ) {
+    }
+
+    public function closure(FlowContext $context) : void
+    {
+        $this->loadedRows = 0;
+        $this->columnsCount = null;
+    }
+
+    public function load(Rows $rows, FlowContext $context) : void
+    {
+        if (!$rows->count()) {
+            return;
+        }
+
+        if (null === $this->columnsCount) {
+            $this->columnsCount = $rows->first()->entries()->count();
+        }
+
+        $currentRow = $this->calculateCurrentRow($rows);
+        $sheetRange = $this->calculateSheetAddress($currentRow);
+
+        $normalizer = new RowsNormalizer(
+            new EntryNormalizer($context->config->caster(), $this->dateTimeFormat, $this->dateFormat, $this->timeFormat)
+        );
+
+        $this->service->spreadsheets_values->update(
+            $this->spreadsheetId,
+            $sheetRange->toString(),
+            new ValueRange(
+                [
+                    'values' => $this->values($rows, $normalizer),
+                    'majorDimension' => 'ROWS',
+                ]
+            ),
+            $this->inputOption->toArray()
+        );
+
+        $this->loadedRows = $currentRow;
+    }
+
+    public function withDateFormat(string $dateFormat) : self
+    {
+        $this->dateFormat = $dateFormat;
+
+        return $this;
+    }
+
+    public function withDateTimeFormat(string $dateTimeFormat) : self
+    {
+        $this->dateTimeFormat = $dateTimeFormat;
+
+        return $this;
+    }
+
+    public function withHeader(bool $withHeader) : self
+    {
+        $this->withHeader = $withHeader;
+
+        return $this;
+    }
+
+    public function withInputOption(ValueInputOption $inputOption) : self
+    {
+        $this->inputOption = $inputOption;
+
+        return $this;
+    }
+
+    public function withTimeFormat(string $timeFormat) : self
+    {
+        $this->timeFormat = $timeFormat;
+
+        return $this;
+    }
+
+    private function calculateCurrentRow(Rows $rows) : int
+    {
+        return $rows->count() + (0 === $this->loadedRows ? ($this->withHeader ? 1 : 0) : $this->loadedRows);
+    }
+
+    private function calculateSheetAddress(int $currentRow) : SheetAddress
+    {
+        return SheetAddress::calculate(
+            (int) $this->columnsCount,
```

Comment:
Shouldn't this be taken from the Rows that are currently being loaded? 
Currently this value comes from the first row, but there is no guarantee that another row will not add more columns.